src/solvers/dgsem_structured/dg_3d.jl

```diff
@@ -263,6 +263,84 @@ end
   end
 end
 
+# Computing the normal vector for the FV method on curvilinear subcells. 
+# To fulfill Freestream-Preservation we use the explicit formula B.53 in Appendix B.4 
+# by Hennemann, Rueda-Ramirez, Hindenlang, Gassner (2020)
+# "A provably entropy stable subcell shock capturing approach for high order split form DG for the compressible Euler Equations"
+# [arXiv: 2008.12044v2] (https://arxiv.org/pdf/2008.12044)
+@inline function calcflux_fv!(fstar1, fstar2, fstar3, u::AbstractArray{<:Any,5},
+                              mesh::Union{StructuredMesh{3}, P4estMesh{3}}, nonconservative_terms::Val{false},
+                              equations, volume_flux_fv, dg::DGSEM, element, cache)
+  @unpack contravariant_vectors = cache.elements
+  @unpack weights, derivative_matrix = dg.basis
+
+  fstar1[:, 1,            :, :] .= zero(eltype(fstar1))
+  fstar1[:, nnodes(dg)+1, :, :] .= zero(eltype(fstar1))
+
+  for k in eachnode(dg), j in eachnode(dg)
+    normal_direction = get_contravariant_vector(1, contravariant_vectors, 1, j, k, element)
+
+    for i in 2:nnodes(dg)
+      u_ll = get_node_vars(u, equations, dg, i-1, j, k, element)
+      u_rr = get_node_vars(u, equations, dg, i,   j, k, element)
+
+      for ii in 1:nnodes(dg)
+        normal_direction += weights[i-1] * derivative_matrix[i-1, ii] * get_contravariant_vector(1, contravariant_vectors, ii, j, k, element)
+      end
```

Comment:
You are talking about the following equation, correct?
![image](https://user-images.githubusercontent.com/12693098/126779095-f3017d16-8d8a-48d0-9648-4cd971afc36f.png)

I would need to read the paper more carefully to be sure, but can you please explain why there is a double sum in the paper while you use only a single sum? And which equation for the FV stuff are you using? Please add a few more comments to make it easier to follow this implementation (same applies to 2D, of course).

Reply:
I use the last equation here. Example for 2 nodes/weights:

What we do for n_{1, j, k}:

n_{1, j, k} = Ja_{1, j, k} + w(1) * D(1, 1) * Ja_{1, j, k}
                                     + w(2) * D(1, 2) * Ja_{2, j, k}
                                     + w(1) * D(2, 1) * Ja_{1, j, k}
                                     + w(2) * D(2, 2) * Ja_{2, j, k}

What the Paper says for n_{1, j, k}:

n_{1, j, k} = Ja_{1, j, k} + Q(1, 1) * Ja_{1, j, k}
                                     + Q(1, 2) * Ja_{2, j, k}
                                     + Q(2, 1) * Ja_{1, j, k}
                                     + Q(2, 2) * Ja_{2, j, k}

whereby Q = M * D, which means that Q(1, 1) = w(1) * D(1, 1) , Q(1, 2) = w(1) * D(1, 2) , Q(2, 1) = w(2) * D(2, 1) , Q(2, 2) = w(2) * D(2, 2).

Therefore it’s equivalent to our approach.

Reply:
I received an email that you wrote something here, @Cczernik, but I cannot see anything - is that correct?

Reply:
I use the last equation here, but I think I just found a mistake in my loops. Example for 2 Nodes/Weights (shifted indices)

What I do for n_{1, j, k}:
```
n_{1, j, k} = Ja_{1, j, k} + w(1) * D(1, 1) * Ja_{1, j, k}
                                     + w(2) * D(1, 2) * Ja_{2, j, k}
                                     + w(1) * D(2, 1) * Ja_{1, j, k}
                                     + w(2) * D(2, 2) * Ja_{2, j, k}
```
                                     
What the Paper says for n_{1, j, k}: 
```
n_{1, j, k} = Ja_{1, j, k} + Q(1, 1) * Ja_{1, j, k}
                                     + Q(1, 2) * Ja_{2, j, k}
                                     (+ Q(2, 1) * Ja_{1, j, k}
                                     + Q(2, 2) * Ja_{2, j, k}) <- The first sum `\sum_{l=0}^{i}` is only true for i = NNodes
```
whereby Q = M * D, which means that Q(1, 1) = w(1) * D(1, 1) , Q(1, 2) = w(1) * D(1, 2) , Q(2, 1) = w(2) * D(2, 1) , Q(2, 2) = w(2) * D(2, 2).

I will check that more carefully and ask a coauthor to look over it.

Reply:
At second glance, the implementation should be correct after all.